README.md
# lua-resty-session

**lua-resty-session** is a secure, and flexible session library for OpenResty.

## Hello World with lua-resty-session

```nginx
worker_processes  1;

events {
    worker_connections  1024;
}

http {
    server {
        listen       8080;
        server_name  localhost;
        default_type text/html;
        location / {
            content_by_lua '
                ngx.say("<html><body><a href=/start>Start the test</a>!</body></html>")
            ';
        }
        location /start {
            content_by_lua '
                local session = require "resty.session".start()
                session.data.name = "OpenResty Fan"
                session:save()
                ngx.say("<html><body>Session started. ",
                        "<a href=/test>Check if it is working</a>!</body></html>")
            ';
        }
        location /test {
            content_by_lua '
                local session = require "resty.session".open()
                ngx.say("<html><body>Session was started by <strong>",
                        session.data.name or "Anonymous",
                        "</strong>! <a href=/destroy>Destroy the session</a>.</body></html>")
            ';
        }
        location /destroy {
            content_by_lua '
                local session = require "resty.session".start()
                session:destroy()
                ngx.say("<html><body>Session was destroyed. ",
                        "<a href=/check>Is it really so</a>?</body></html>")
            ';
        }
        location /check {
            content_by_lua '
                local session = require "resty.session".open()
                ngx.say("<html><body>Session was really destroyed, you are known as ",
                        "<strong>",
                        session.data.name or "Anonymous",
                        "</strong>! <a href=/>Start again</a>.</body></html>")
            ';
        }
    }
}
```

## Roadmap

* [ ] Add support for hmac plugins (e.g. in case where we have cipher that already supports authentication with associated data aka AEAD)
* [ ] Add support for session id generator plugins (maybe you don't want to use random data, and want UUID or maybe some kind of database identifier instead).
* [ ] Add support for `lua-resty-nettle` for more wide variety of encryption algorithms as a plugin.
* [ ] Implement cookieless server-side session support using `ssl_session_id` as a `session.id` (using a server-side storage).
* [x] Implement `SameSite` cookie flag [DONE].

## Installation

Just place [`session.lua`](https://github.com/bungle/lua-resty-session/blob/master/lib/resty/session.lua) and [`session`](https://github.com/bungle/lua-resty-session/tree/master/lib/resty/session) directory somewhere in your `package.path`, preferably under `resty` directory. If you are using OpenResty, the default location would be `/usr/local/openresty/lualib/resty`.

### Using LuaRocks or MoonRocks

If you are using LuaRocks >= 2.2:

```Shell
$ luarocks install lua-resty-session
```

If you are using LuaRocks < 2.2:

```Shell
$ luarocks install --server=http://rocks.moonscript.org moonrocks
$ moonrocks install lua-resty-session
```

MoonRocks repository for `lua-resty-session` is located here: https://rocks.moonscript.org/modules/bungle/lua-resty-session.

## About The Defaults

`lua-resty-session` does by default set session only cookies (non-persistent, and `HttpOnly`) so that
the cookies are not readable from Javascript (not subjectible to XSS in that matter). It will also set
`Secure` flag by default when the request was made via SSL/TLS connection. Cookies send via SSL/TLS
don't work when sent via HTTP and vice-versa (unless the checks are disabled). By default the HMAC key
is generated from session id (random bytes generated with OpenSSL), expiration time, unencrypted
data, and Nginx variables `ssl_session_id` (if requested with TLS/SSL), `http_user_agent` and `scheme`.
You may also configure it to use `remote_addr` as well by setting `set $session_check_addr on;`
(but this may be problematic with clients behind proxies or NATs that change the remote address
between requests).

The data part is encrypted with AES-algorithm (by default it uses OpenSSL `EVP_aes_256_cbc` and
`EVP_sha512` functions that are provided with `lua-resty-string`. They come pre-installed with
the default OpenResty bundle. The `lua-resty-session` library is not tested with all the
`resty.aes` functions (but the defaults are tested to be working). Please let me know or contact
`lua-resty-string` project if you hit any problems with different algorithms. We also support
pluggable cipher adapters. You can also disable encryption by choosing `none` adapter.

Session identifier length is by default 16 bytes (randomly generated data with OpenSSL
`RAND_bytes` function). The server secret is also generated by default with this same
function and it's default length is 32 bytes. This will work until Nginx is restarted, but you
might want to consider setting your own secret using `set $session_secret 623q4hR325t36VsCD3g567922IC0073T;`,
for example (this will work in farms installations as well, but you are then responsible for
rotating the secret). On farm installations you should also configure other session configuration
variables the same on all the servers in the farm.

Cookie parts are encoded with cookie safe Base64 encoding (we also support pluggable encoders).
Before encrypting and encoding the data part, the data is serialized with JSON encoding (so you can
use basic Lua types in data, and expect to receive them back as the same Lua types). JSON encoding
is done by the bundled OpenResty cJSON library (Lua cJSON). We do support pluggable serializers as
well, though only serializer currently supplied is JSON. Cookie's path scope is by default `/`
(meaning that it will be send to all paths in the server. The domain scope is not set by default,
and it means that the cookie will only be sent back to same domain where it originated.

For session data we do support pluggable storage adapters. The default adapter is `cookie` that
stores data to client-side cookie. Currently we do also support a few server side storages: `shm`
(aka a shared dictionary), `memcache`, and `redis`.

## Notes About Turning Lua Code Cache Off

In issue ([#15](https://github.com/bungle/lua-resty-session/issues/15)) it was raised that there may
be problems of using `lua-resty-session` when the `lua_code_cache` setting has been turned off.

Nginx:

```nginx
lua_code_cache off;
```

The problem is caused by the fact that by default we do generate session secret automatically with
a random generator (on first use of the library). If the code cache is turned off, we regenerate
the secret on each request. That will invalidate the cookies aka making sessions non-functioning.
The cure for this problem is to define the secret in Nginx or in Lua code (it is a good idea to
always have session secret defined).

Nginx:

```nginx
set $session_secret 623q4hR325t36VsCD3g567922IC0073T;
```

Lua:

```lua
local session = require "resty.session".start{ secret = "623q4hR325t36VsCD3g567922IC0073T" }
-- or
local session = require "resty.session".new()
session.secret = "623q4hR325t36VsCD3g567922IC0073T"
```

## Pluggable Storage Adapters

With version 2.0 we started to support pluggable session data storage adapters. We do currently have
support for these backends:

* `cookie` aka Client Side Cookie (this is the default adapter)
* `shm` aka Lua Shared Dictionary
* `memcache` aka Memcached Storage Backend (thanks [@zandbelt](https://github.com/zandbelt))
* `redis` aka Redis Backend

Here are some comparisons about the backends:

|                               | cookie | shm  | memcache | redis |
| :---------------------------- | :----: | :--: | :------: | :---: |
| Stateless                     | ✓      |      |          |       |
| Lockless                      | ✓      | ¹    | ¹        | ¹     |
| Works with Web Farms          | ✓      |      | ✓        | ✓     |
| Session Data Stored on Client | ✓      |      |          |       |
| Zero Configuration            | ✓      |      |          |       |
| Extra Dependencies            |        |      | ✓        | ✓     |
| Extra Security ²              |        | ✓    | ✓        | ✓     | 

¹ Can be configured lockless.

² HMAC is stored on a client but the data is stored on a server. That means that you are unable to edit
  cookie if you cannot edit server side storage as well, and vice-versa.

The storage adapter can be selected from Nginx config like this:

```nginx
set $session_storage shm;
```

Or with Lua code like this:

```lua
local session = require "resty.session".new()
-- After new you cannot specify storage as a string, you need to give actual implementation
session.storage = require "resty.sesstion.storage.shm"
-- or
local session = require "resty.session".new({ storage = "shm" })
```

#### Cookie Storage Adapter

Cookie storage adapter is the default adapter that is used if storage adapter has not been configured. Cookie
adapter does not have any settings.

Cookie adapter can be selected with configuration (if no configuration is present, the cookie adapter is picked up):

```nginx
set $session_storage cookie;
```

#### Shared Dictionary Storage Adapter

Shared dictionary uses OpenResty shared dictionary and works with multiple worker processes, but it isn't a good
choice if you want to run multiple separate frontends. It is relatively easy to configure and has some added
benefits on security side compared to `cookie`, although the normal cookie adapter is quite secure as well.
For locking the `shm` adapter uses `lua-resty-lock`.

Shared dictionary adapter can be selected with configuration:

```nginx
set $session_storage shm;
```

But for this to work, you will also need a storage configured for that:

```nginx
http {
   lua_shared_dict sessions 10m;
}
```

Additionally you can configure the locking and some other things as well:

```nginx
set $session_shm_store         sessions;
set $session_shm_uselocking    on;
set $session_shm_lock_exptime  30;
set $session_shm_lock_timeout  5;
set $session_shm_lock_step     0.001;
set $session_shm_lock_ratio    2;
set $session_shm_lock_max_step 0.5;
```

The keys stored in shared dictionary are in form:

`{session id}` and `{session id}.lock`.


#### Memcache Storage Adapter

Memcache storage adapter stores the session data inside Memcached server.
It is scalable and works with web farms. 

Memcache adapter can be selected with configuration:

```nginx
set $session_storage memcache;
```

Additionally you can configure Memcache adapter with these settings:

```nginx
set $session_memcache_prefix        sessions;
set $session_memcache_socket        unix:///var/run/memcached/memcached.sock;
set $session_memcache_host          127.0.0.1;
set $session_memcache_port          11211;
set $session_memcache_uselocking    on;
set $session_memcache_spinlockwait  10000;
set $session_memcache_maxlockwait   30;
set $session_memcache_pool_timeout  45;
set $session_memcache_pool_size     10;
```

The keys stored in Memcached are in form:

`{prefix}:{session id}` and `{prefix}:{session id}.lock`.

#### Redis Storage Adapter

Redis storage adapter stores the session data inside Redis server.
It is scalable and works with web farms. 

Redis adapter can be selected with configuration:

```nginx
set $session_storage redis;
```

Additionally you can configure Redis adapter with these settings:

```nginx
set $session_redis_prefix        sessions;
set $session_redis_socket        unix:///var/run/redis/redis.sock;
set $session_redis_host          127.0.0.1;
set $session_redis_port          6379;
set $session_redis_uselocking    on;
set $session_redis_spinlockwait  10000;
set $session_redis_maxlockwait   30;
set $session_redis_pool_timeout  45;
set $session_redis_pool_size     10;
```

The keys stored in Redis are in form:

`{prefix}:{session id}` and `{prefix}:{session id}.lock`.

#### Implementing a Storage Adapter

It is possible to implement additional storage adapters using the plugin architecture in `lua-resty-session`.

You need to implement at least these APIs:

* `table adapter.new(opts)`
* `id, expires, data, hmac adapter:open(cookie, lifetime)`
* `cookie adapter:save(id, expires, data, hmac)`

There are a few additional hooks that you may want to attach:

* `ok, error adapter:start(id)`
* `ok, error adapter:destroy(id)`

You have to place your adapter inside `resty.session.storage` for auto-loader to work.

To configure session to use your adapter, you can do so with Nginx configuration (or in Lua code):

```nginx
# Just an example. Pull request for MySQL support are greatly welcomed.
set $session_storage mysql;
```

## Pluggable Ciphers

With version 2.1 we started to support pluggable ciphers. We currently have support for these ciphers:

* `aes` aka AES encryption / decryption using `lua-resty-string`'s AES library (the default).
* `none` aka no encryption or decryption is done.

The cipher adapter can be selected from Nginx config like this:

```nginx
set $session_cipher aes;
```

Or with Lua code like this:

```lua
local session = require "resty.session".start{ cipher = "aes" }
```

#### AES Cipher

AES Cipher uses `lua-resty-string`'s (an OpenResty core library) AES implementation
(bindings to OpenSSL) for encryption.

AES adapter can be selected with configuration:

```nginx
set $session_cipher aes;
```

Additionally you can configure Memcache adapter with these settings:

```nginx
set $session_aes_size   256;
set $session_aes_mode   "cbc";
set $session_aes_hash   "sha512";
set $session_aes_rounds 1;
```

Here follows the description of each setting:

**size**

`session.aes.size` holds the size of the cipher (`lua-resty-string` supports AES in `128`, `192`,
and `256` bits key sizes). See `aes.cipher` function in `lua-resty-string` for more information.
By default this will use `256` bits key size. This can be configured with Nginx
`set $session_aes_size 256;`.

**mode**

`session.aes.mode` holds the mode of the cipher. `lua-resty-string` supports AES in `ecb`, `cbc`,
`cfb1`, `cfb8`, `cfb128`, `ofb`, and `ctr` modes (ctr mode is not available with 256 bit keys).
See `aes.cipher` function in `lua-resty-string` for more information. By default `cbc` mode is
used. This can be configured with Nginx `set $session_aes_mode cbc;`.

**hash**

`session.aes.hash` is used in ecryption key, and iv derivation (see: OpenSSL
[EVP_BytesToKey](https://www.openssl.org/docs/crypto/EVP_BytesToKey.html)). By default `sha512` is
used but `md5`, `sha1`, `sha224`, `sha256`, and `sha384` are supported as well in `lua-resty-string`.
This can be configured with Nginx `set $session_aes_hash sha512;`.

**rounds**

`session.aes.rounds` can be used to slow-down the encryption key, and iv derivation. By default
this is set to `1` (the fastest). This can be configured with Nginx `set $session_aes_rounds 1;`.

#### None Cipher

None cipher disables encryption of the session data. This can be handy if you want to
debug things or want you session management as light as possible, or perhaps share the
session data with some other process without having to deal with encryption key management.
In general it is better to have encryption enabled in a production.

None adapter can be selected with configuration:

```nginx
set $session_cipher none;
```

There isn't any settings for None adapter as it is basically a no-op adapter.

#### Implementing a Cipher Adapter

If you want to write your own cipher adapter, you need to implement these three methods:

* `table adapter.new(opts)`
* `string adapter:encrypt(data, key, iv or salt, associated data)`
* `string adapter:decrypt(ciphertext, key, iv or salt, associated data)`

If you do not use say iv or associated data in your cipher, you can ignore them.

You have to place your adapter inside `resty.session.ciphers` for auto-loader to work.

## Pluggable Serializers

Currently we only support JSON serializer, but there is a plugin architecture that you can use to
plugin your own serializer. The serializer is used to serialize session data in a form that can be
later deserialized and stored in some of our supported storages.

The supported serializer names are:

* `json`

You need only to implement two functions to write an adapter:

* `string adapter.serialize(table)`
* `table adapter.deserialize(string)`

You have to place your adapter inside `resty.session.serializers` for auto-loader to work.

To configure session to use your adapter, you can do so with Nginx configuration (or in Lua code):

```nginx
set $session_serializer json;
```

## Pluggable Encoders

Cookie data needs to be encoded in cookie form before it is send to client. We support
two encoding methods by default: modified cookie friendly base-64, and base-16 (or hexadecimal encoding).

The supported encoder names are:

* `base64`
* `base16` or `hex`

If you want to write your own encoder, you need to implement these two methods:

* `string adapter.encode(string)`
* `string adapter.decode(string)`

You have to place your adapter inside `resty.session.encoders` for auto-loader to work.

To configure session to use your adapter, you can do so with Nginx configuration (or in Lua code):

```nginx
set $session_encoder base64;
```

## Lua API

### Functions and Methods

#### table session.new(opts or nil)

With this function you can create a new session table (i.e. the actual session instance). This allows
you to generate session table first, and set invidual configuration before calling `session:open()` or
`session:start()`. You can also pass in `opts` Lua `table` with the configurations.

```lua
local session = require "resty.session".new()
-- set the configuration parameters before calling start
session.cookie.domain = ".mydomain.com"
-- call start before setting session.data parameters
session:start()
session.data.uid = 1
-- save session and update the cookie to be sent to the client
session:save()
```

This is equivalent to this:

```lua
local session = require "resty.session".new{ cookie = { domain = ".mydomain.com" } }
session:start()
session.data.uid = 1
session:save()
```

As well as with this:

```lua
local session = require "resty.session".start{ cookie = { domain = ".mydomain.com" } }
session.data.uid = 1
session:save()
```

#### table, boolean session.open(opts or nil)

With this function you can open a new session. It will create a new session Lua `table` on each call (unless called with
colon `:` as in examples above with `session.new`). Calling this function repeatedly will be a no-op when using colon `:`.
This function will return a (new) session `table` as a result. If the session cookie is supplied with user's HTTP(S)
client then this function validates the supplied session cookie. If validation is successful, the user supplied session
data will be used (if not, a new session is generated with empty data). You may supply optional session configuration
variables with `opts` argument, but be aware that many of these will only have effect if the session is a fresh session
(i.e. not loaded from user supplied cookie). The second `boolean` return argument will be `true` if the user
client send a valid cookie (meaning that session was already started on some earlier request), and `false` if the
new session was created (either because user client didn't send a cookie or that the cookie was not a valid one). This
function will not set a client cookie. You need to call `session:start()` to really start the session. This open function
is mainly used if you only want to read data and avoid automatically sending a cookie (see also issue
[#12](https://github.com/bungle/lua-resty-session/issues/12)). But be aware that this doesn't update cookie
expiration time stored in a cookie.

```lua
local session = require "resty.session".open()
-- Set some options (overwriting the defaults or nginx configuration variables)
local session = require "resty.session".open{ identifier = { length = 32 }}
-- Read some data
if session.present then
    ngx.print(session.data.uid)
end
-- Now let's really start the session
-- (session.started will be always false in this example):
if not session.started then 
    session:start()
end
session.data.greeting = "Hello, World!"
session:save()
```

#### table, boolean session.start(opts or nil)


CHANGES
Changes with lua-resty-session 2.9                                1 Sep 2016

    *) Bugfix: Weird bug where RAND_bytes was not working on Windows
       platform. Code changed to use resty.random.

       See Also: https://github.com/bungle/lua-resty-session/issues/31

       Thanks @gtuxyco

Changes with lua-resty-session 2.8                                5 Jul 2016

    *) Bugfix: AES Cipher used a wrong table for cipher sizes.

       See Also: https://github.com/bungle/lua-resty-session/issues/30

       Thanks @pronan

Changes with lua-resty-session 2.7                               18 May 2016

    *) Feature: Redis storage adapter now supports Redis authentication.

       See Also: https://github.com/bungle/lua-resty-session/pull/28

       Thanks @cheng5533062

Changes with lua-resty-session 2.6                               18 Apr 2016

    *) Just cleanups and changed _VERSION to point correct version.

Changes with lua-resty-session 2.5                               18 Apr 2016

    *) Bugfix: session.save close argument was not defaulting to true.

Changes with lua-resty-session 2.4                               17 Apr 2016

    *) Feature: Cookie will now have SameSite attribute set as "Lax" by
       default. You can turn it off or set to "Strict" by configuration.

    *) Change: Calling save will now also set session.id if the save
       was called without calling start first.

       See Also: https://github.com/bungle/lua-resty-session/issues/27

       Thanks @hcaihao

Changes with lua-resty-session 2.3                               16 Oct 2015

    *) Bugfix: Fixes issue #19 where regenerating session would
       throw an error when using cookie storage.

       See Also: https://github.com/bungle/lua-resty-session/issues/19

       Thanks @hulu1522

Changes with lua-resty-session 2.2                               17 Sep 2015

    *) Change: Removed all session_cipher_* deprecated settings
       (it was somewhat broken in 2.1).

    *) Change: Changed session secret to be by default 32 bytes random data

    See Also: https://github.com/bungle/lua-resty-session/issues/18

    Thanks @iain-buclaw-sociomantic

    *) Documentation: Added documentation about removed features and
       corrected about session secret size accordingly.

Changes with lua-resty-session 2.1                                7 Sep 2015

    *) Feature: Added architecture for Cipher adapter plugins.

    See Also: https://github.com/bungle/lua-resty-session/issues/16

    Thanks @mingfang

    *) Feature: Implemented AES cipher adapter (just like it was before)

    *) Feature: Implemented None cipher adapter (no encryption)

    *) Change: Changed JSON serializer to use cjson.safe instead

    *) Documentation: Added documentation about pluggable ciphers

Changes with lua-resty-session 2.0                               31 Aug 2015

    *) Feature: Added architecture for Storage adapter plugins.

       See Also: https://github.com/bungle/lua-resty-session/issues/13

    *) Feature: Implemented Client Side Cookie storage adapter.

    *) Feature: Implemented Memcache storage adapter.

       See Also: https://github.com/bungle/lua-resty-session/pull/14

       Thanks @zandbelt

    *) Feature: Implemented Redis storage adapter.

    *) Feature: Implemented Shared Dictionary (shm) storage adapter.

    *) Feature: Added architecture for Encoder and Decoder plugins.

    *) Feature: Implemented Base 64 encoder / decoder.

    *) Feature: Implemented Base 16 (hex) encoder / decoder.

    *) Feature: Added architecture for Serializer plugins

    *) Feature: Implemented JSON serializer.

    *) Feature: Persistent cookies will now also contain Max-Age in addition
       to Expires.

    *) Change: Cookie domain attribute is not set anymore if not specified.

    *) Documentation: Added notes about using lua-resty-session with
       Lua code cache turned off.

       See also: https://github.com/bungle/lua-resty-session/issues/15

       Thanks @BizShuk

Changes with lua-resty-session 1.7                                3 Aug 2015

    *) Feature: Added session.open() function that only opens a session
       but doesn't send the cookie (until start is called).

       See also: https://github.com/bungle/lua-resty-session/issues/12

       Thanks @junhanamaki

    *) Bugfix: Fixed cookie expiration time format on Firefox bug:
       https://github.com/bungle/lua-resty-session/pull/10

       Thanks @junhanamaki

    *) Bugfix: Fixed an issue of overwriting a variable:
       https://github.com/bungle/lua-resty-session/pull/11

       Thanks @junhanamaki

Changes with lua-resty-session 1.6                                5 May 2015

    *) Bugfix: Fixed truncated cookie value bug:
       https://github.com/bungle/lua-resty-session/pull/8

       Thanks @kipras

Changes with lua-resty-session 1.5                               27 Nov 2014

    *) Bugfix: Cookies are not always "secure":
       https://github.com/bungle/lua-resty-session/issues/5

       Thanks @vladimir-smirnov-sociomantic

    *) Documentation: Added documentation about Nginx SSL/TLS configuration
       settings related to session lifetime and ssl session ids.

Changes with lua-resty-session 1.4                               26 Nov 2014

    *) Bugfix: Fixed an issue where session configurations did get cached
       on a module level. This issue is discussed in pull-request #4:
       https://github.com/bungle/lua-resty-session/pull/4

       Thanks @kipras.

    *) Feature: Added session.new function.

    *) Change: session.start{ ... } (a call with config parameters) works
       now as expected.

    *) Change: session.start now returns additional extra boolean parameter
       that can be used to check if the session is s new session (false) or
       s previously started one (true).

    *) Documentation: Added documentation about Nginx configuration used

lib/resty/session.lua
local require      = require
local var          = ngx.var
local header       = ngx.header
local concat       = table.concat
local hmac         = ngx.hmac_sha1
local time         = ngx.time
local http_time    = ngx.http_time
local find         = string.find
local type         = type
local pcall        = pcall
local tonumber     = tonumber
local setmetatable = setmetatable
local getmetatable = getmetatable
local random       = require "resty.random".bytes

local function enabled(val)
    if val == nil then return nil end
    return val == true or (val == "1" or val == "true" or val == "on")
end

local function setcookie(session, value, expires)
    local c = session.cookie
    local i = 3
    local n = session.name .. "="
    local k = { n, value or "" }
    local d = c.domain
    local x = c.samesite
    if expires then
        k[i] = "; Expires=Thu, 01 Jan 1970 00:00:01 GMT; Max-Age=0"
        i=i+1
    elseif c.persistent then
        k[i]   = "; Expires="
        k[i+1] = http_time(session.expires)
        k[i+2] = "; Max-Age="
        k[i+3] = c.lifetime
        i=i+4
    end
    if d and d ~= "localhost" and d ~= "" then
        k[i]   = "; Domain="
        k[i+1] = d
        i=i+2
    end
    k[i]   = "; Path="
    k[i+1] = c.path or "/"
    i=i+2
    if x == "Lax" or x == "Strict" then
        k[i] = "; SameSite="
        k[i+1] = x
        i=i+2
    end
    if c.secure then
        k[i] = "; Secure"
        i=i+1
    end
    if c.httponly then
        k[i] = "; HttpOnly"
    end
    k = concat(k)
    local s = header["Set-Cookie"]
    local t = type(s)
    if t == "table" then
        local f = false
        local z = #s
        for i=1, z do
            if find(s[i], n, 1, true) == 1 then
                s[i] = k
                f = true
                break
            end
        end
        if not f then
            s[z+1] = k
        end
    elseif t == "string" and find(s, n, 1, true) ~= 1  then
        s = { s, k }
    else
        s = k
    end
    header["Set-Cookie"] = s
    return true
end

local function save(session, close)
    session.expires = time() + session.cookie.lifetime
    local i, e, s = session.id, session.expires, session.storage
    local k = hmac(session.secret, i .. e)
    local d = session.serializer.serialize(session.data)
    local h = hmac(k, concat{ i, e, d, session.key })
    local cookie, err = s:save(i, e, session.cipher:encrypt(d, k, i, session.key), h, close)
    if cookie then
        return setcookie(session, cookie)
    end
    return nil, err
end

local function regenerate(session, flush)
    local i = session.present and session.id or nil
    session.id = random(session.identifier.length, true) or random(session.identifier.length)
    if flush then
        if i and session.storage.destroy then
            session.storage:destroy(i);
        end
        session.data = {}
    end
end

local persistent = enabled(var.session_cookie_persistent or false)
local defaults = {
    name       = var.session_name       or "session",
    storage    = var.session_storage    or "cookie",
    serializer = var.session_serializer or "json",
    encoder    = var.session_encoder    or "base64",
    cipher     = var.session_cipher     or "aes",
    cookie = {
        persistent = persistent,
        renew      = tonumber(var.session_cookie_renew)    or 600,
        lifetime   = tonumber(var.session_cookie_lifetime) or 3600,
        path       = var.session_cookie_path               or "/",
        domain     = var.session_cookie_domain,
        samesite   = var.session_cookie_samesite           or "Lax",
        secure     = enabled(var.session_cookie_secure),
        httponly   = enabled(var.session_cookie_httponly   or true),
        delimiter  = var.session_cookie_delimiter          or "|"
    }, check = {
        ssi    = enabled(var.session_check_ssi    or persistent == false),
        ua     = enabled(var.session_check_ua     or true),
        scheme = enabled(var.session_check_scheme or true),
        addr   = enabled(var.session_check_addr   or false)
    }, identifier = {
        length  = tonumber(var.session_identifier_length) or 16
    }
}
defaults.secret = var.session_secret or random(32, true) or random(32)

local session = {
    _VERSION = "2.9"
}

session.__index = session

function session.new(opts)
    if getmetatable(opts) == session then
        return opts
    end
    local z = defaults
    local y = opts or z
    local a, b = y.cookie     or z.cookie,     z.cookie
    local c, d = y.check      or z.check,      z.check
    local e, f = y.cipher     or z.cipher,     z.cipher
    local g, h = y.identifier or z.identifier, z.identifier
    local o, i = pcall(require, "resty.session.storage." .. (y.storage or z.storage))
    if not o then
        i = require "resty.session.storage.cookie"
    end
    local o, j = pcall(require, "resty.session.serializers." .. (y.serializer or z.serializer))
    if not o then
        j = require "resty.session.serializers.json"
    end
    local o, k = pcall(require, "resty.session.encoders." .. (y.encoder or z.encoder))
    if not o then
        k = require "resty.session.encoders.base64"
    end
    local o, l = pcall(require, "resty.session.ciphers." .. (e or f))
    if not o then
        l = require "resty.session.ciphers.aes"
    end
    local self = {
        name       = y.name    or z.name,
        serializer = j,
        encoder    = k,
        data       = y.data    or {},
        secret     = y.secret  or z.secret,
        cookie = {
            persistent = a.persistent or b.persistent,
            renew      = a.renew      or b.renew,
            lifetime   = a.lifetime   or b.lifetime,
            path       = a.path       or b.path,
            domain     = a.domain     or b.domain,
            samesite   = a.samesite   or b.samesite,
            secure     = a.secure     or b.secure,
            httponly   = a.httponly   or b.httponly,
            delimiter  = a.delimiter  or b.delimiter
        }, check = {
            ssi        = c.ssi        or d.ssi,
            ua         = c.ua         or d.ua,
            scheme     = c.scheme     or d.scheme,
            addr       = c.addr       or d.addr
        }, identifier = {
            length     = g.length     or h.length
        }
    }
    self.storage = i.new(self)
    self.cipher = l.new(self)
    return setmetatable(self, session)
end

function session.open(opts)
    local self = opts
    if getmetatable(self) == session then
        if self.opened then
            return self, self.present
        end
    else
        self = session.new(opts)
    end
    local scheme = header["X-Forwarded-Proto"]
    if self.cookie.secure == nil then
        if scheme then
            self.cookie.secure = scheme == "https"
        else
            self.cookie.secure = var.https == "on"
        end
    end
    scheme = self.check.scheme and (scheme or var.scheme or "") or ""
    local addr = ""
    if self.check.addr then
        addr = header["CF-Connecting-IP"] or
               header["Fastly-Client-IP"] or
               header["Incap-Client-IP"]  or
               header["X-Real-IP"]
        if not addr then
            addr = header["X-Forwarded-For"]
            if addr then
                -- We shouldn't really get the left-most address, because of spoofing,
                -- but this is better handled with a module, like nginx realip module,
                -- anyway (see also: http://goo.gl/Z6u2oR).
                local s = find(addr, ',', 1, true)
                if s then
                    addr = addr:sub(1, s - 1)
                end
            else
                addr = var.remote_addr
            end
        end
    end
    self.key = concat{
        self.check.ssi and (var.ssl_session_id  or "") or "",
        self.check.ua  and (var.http_user_agent or "") or "",
        addr,
        scheme
    }
    local cookie = var["cookie_" .. self.name]
    if cookie then
        local i, e, d, h = self.storage:open(cookie, self.cookie.lifetime)
        if i and e and e > time() and d and h then
            self.id = i
            self.expires = e
            local k = hmac(self.secret, self.id .. e)
            d = self.cipher:decrypt(d, k, i, self.key)
            if d and hmac(k, concat{ i, e, d, self.key }) == h then
                self.data = self.serializer.deserialize(d)
                self.present = true
            end
        end
    end
    if not self.present then
        regenerate(self)
    end
    if type(self.data) ~= "table" then self.data = {} end
    self.opened = true
    return self, self.present
end

function session.start(opts)
    if getmetatable(opts) == session and opts.started then
        return opts, opts.present
    end
    local self, present = session.open(opts)
    if present then
        if self.storage.start then
            self.storage:start(self.id)
        end
        if self.expires - time() < self.cookie.renew then
            save(self)
        end
    else
        save(self)
    end
    self.started = true
    return self, present
end

function session:regenerate(flush)
    regenerate(self, flush)
    return save(self)
end

function session:save(close)
    if not self.id then
        self.id = random(self.identifier.length, true) or random(self.identifier.length)
    end
    return save(self, close ~= false)
end

function session:destroy()
    if self.storage.destroy then
        self.storage:destroy(self.id)
    end
    self.data      = {}
    self.present   = nil
    self.opened    = nil
    self.started   = nil
    self.destroyed = true
    return setcookie(self, "", true)
end

return session